File: intro_lab/solutions/pi_loop.f90
! This program computes pi as
! \pi = 4 arctan(1)
!     = 4 \int _0 ^1 \frac{1} {1 + x^2} dx

program compute_pi

use omp_lib

implicit none

integer(kind=8), parameter :: NSTEPS = 134217728
integer, parameter :: MAXTHREADS = 4

integer(kind=8) :: i
integer :: nthreads
real(kind=8) :: dx, x, pi, ref_pi
real(kind=8) start_time, run_time

dx = 1.0D0 / NSTEPS

do nthreads = 1, MAXTHREADS

    pi = 0.0D0

    call OMP_SET_NUM_THREADS(nthreads)
    start_time = OMP_GET_WTIME()

    !$omp parallel

        !$omp single
            print '("nthreads = ", i0)', nthreads
        !$omp end single
           
        !$omp do private(x) reduction(+:pi)
            do i = 1, NSTEPS
                x = (i - 0.5D0) * dx
                pi = pi + 1.0D0 / (1.0D0 + x * x)
            enddo
        !$omp end do

    !$omp end parallel

    pi = pi * 4.0D0 * dx
    run_time = OMP_GET_WTIME() - start_time
    ref_pi = 4.0D0 * atan(1.0D0)
    print '("pi with ", i0, " steps is ", f16.10, " in ", f12.6, " seconds (error=", e12.6, ")")', NSTEPS, pi, run_time, abs(ref_pi - pi)

enddo

end program

File: intro_lab/solutions/pi_loop.c
#include <stdio.h>
#include <math.h>
#include <omp.h>

#define NSTEPS 134217728
#define MAXTHREADS 4

/*
 * This program computes pi as
 * \pi = 4 arctan(1)
 *     = 4 \int _0 ^1 \frac{1} {1 + x^2} dx
 */
int main(int argc, char** argv)
{
    double dx = 1.0 / NSTEPS;

    int nthreads;
    for (nthreads = 1; nthreads <= MAXTHREADS; nthreads++)
    {
        long i;
        double x, pi = 0.0;

        omp_set_num_threads(nthreads);
        double start_time = omp_get_wtime();

        #pragma omp parallel  
        {
            #pragma omp single
            printf("num_threads = %d,  ", nthreads);

            #pragma omp for private(x) reduction(+:pi)
            for (i = 0; i < NSTEPS; i++)
            {
                x = (i + 0.5) * dx;
                pi += 1.0 / (1.0 + x * x);
            }
        }

        pi *= 4.0 * dx;
        double run_time = omp_get_wtime() - start_time;
        double ref_pi = 4.0 * atan(1.0);
        printf("pi with %ld steps is %.10f in %.6f seconds (error=%e)\n",
               NSTEPS, pi, run_time, fabs(ref_pi - pi));
    }

    return 0;
}
